### PR TITLE
[breaking] Add err chan to log producer and don't panic on error

### PR DESCRIPTION
**Pull request by Tofel**
## What does this PR do?

As per [conversation](https://github.com/testcontainers/testcontainers-go/pull/1278#issuecomment-1784715136) I changed the way starting of log producer works. Now it won't panic if there's an error and instead send it over an error channel. When that happens producer state is cleaned, so that it can be started again. Callers can listen to to the error channel and if there's an error decide how to handle.

Also, it's now required to pass context timeout value, when starting log producer. I've added some boundaries there, it has to be within `<5,60>` seconds. If it's too low the http connection to get the logs is too short and some logs might get missed randomly (happened, when I was setting it to 1s), if it's too long the connection might timeout.

Now... I was thinking about not breaking the method signature and instead adding a new field to the `Container` struct that would store the timeout by default set to 5 seconds and then having a setter method to modify it. If you think that's better I could look into it.

Last, but not least. Tests are very simple ones. In our downstream project I have more complex ones which actually focus on testing a detached goroutine that's listening to errors and restarting log producer if it's needed. If current approach is accepted I could play around with Mockery and add some tests here as well.

## Why is it important?

We don't want unrecoverable panics if there's an error, when reading logs. It should be up to client to decide what should be done if there's an issue.

## Related issues

https://github.com/testcontainers/testcontainers-go/pull/1278
https://github.com/testcontainers/testcontainers-go/pull/1911

## How to test this PR

No straight-forward way. In our downstream project we are using only a limited set of methods from `Container` interface, so I narrowed it even further, put them in a separate interface and in tests created a mocked container with very stripped-down log-related logic that allowed me inject errors at various points. Without it, though... it's not trivial since you'd need to either disrupt/pause the network connection with Docker container so that reading logs fails due to timeout. Or use a bogus container that makes the connection timeout.

### CODE REVIEW

**Comment by mdelapenya** [comment]
Before starting the review of this PR, I'd like to let you know that I'd like to plan a total refactor of the log producers feature: https://github.com/testcontainers/testcontainers-go/issues/525, so would like to know if we could work on this as a solid feature, instead of patching something that is not as usable as desired

**Comment by Tofel** [comment]
@mdelapenya gotcha. Regarding what you wanna do there, we get the logs from log producer and then our log component has different log targets: in-memory, file, loki. It buffers all logs and then flushes then, when you want to.

Anyway, I will leave the PR open, because what you wanna do doesn't sound like something easy and fast to do :-)

File: docker.go
Comment (mdelapenya) [inline docker.go:622]: Wdyt about this: we could probably create functional options for the StartLogProduce method:

```go
func (c *DockerContainer) StartLogProducer(ctx context.Context, opts ...LogProduceOptions) error
```

Being the timeout one of them. That would reduce the breaking change (BC) surface (only people assigning the function to a variable will get impacted). 

I'd try to avoid adding a BC here, but for this upcoming release we already have two (https://github.com/testcontainers/testcontainers-go/releases). So in the end I'm fine adding it if you consider it's fine too :)

**Review by mdelapenya — COMMENTED** [review]
@Tofel I left a few comments before merging this. Please let me know what you think

File: docker.go
Comment (Tofel) [reply docker.go]: Yeah, I was already thinking about that. I think it's a good idea. Applied.

File: docker.go
Comment (Tofel) [reply docker.go:779]: Well, that's the only way for the consumer to be able to listen to errors and react to them. I can point you to our code where we use it, although it might not be the clearest example as there we retry starting/listening in a detached goroutine. 

When I have a second I'd like to use Mockery or something similar to mock
```
r, err := c.provider.client.ContainerLogs(ctx, c.GetContainerID(), options)
```

so that we can cause timeouts or other read errors to make sure the channel works as expected and provide an example of how to use it. WDYT?

File: docker.go
Comment (Tofel) [reply docker.go]: Why not! Done.

File: docker.go
Comment (mdelapenya) [reply docker.go:779]: You already have a `c.Logs(ctx)` method in the container interface. Probably not the simplest thing to have your own mock for that, but it could work

Comment (mdelapenya) [reply docker.go:779]: It's the other way around: `c.Logs` calls `c.provider.client.ContainerLogs`: https://github.com/testcontainers/testcontainers-go/blob/9c97abfa233364d792ff7723606bebc242237c19/docker.go#L323

**Comment by Tofel** [comment]
@mdelapenya is there anything left for me to do before we can merge it? I checked the failure in the pipeline, but it doesn't seem related to my changes at all:
```
=== Failed
=== FAIL: . TestPrependHubRegistrySubstitutor/should_prepend_the_hub_registry_to_images_from_Docker_Hub/plain_image (0.00s)
    image_substitutors_test.go:27: expected my-registry/foo, got foo:latest

=== FAIL: . TestPrependHubRegistrySubstitutor/should_prepend_the_hub_registry_to_images_from_Docker_Hub (0.00s)

=== FAIL: . TestPrependHubRegistrySubstitutor (0.00s)
```

File: logconsumer_test.go
Comment (mdelapenya) [inline logconsumer_test.go:457]: Typo 🙏 
```suggestion
func Test_StartLogProducerStillStartsWithTooLowTimeout(t *testing.T) {
```

**Review by mdelapenya — DISMISSED** [review]
LGTM, thanks!

I added a NIT comment regarding a test method name, but I will merge it as soon it's changed, without waiting for the CI

**Review by mdelapenya — APPROVED** [review]
LGTM! Thanks!

**Comment by mdelapenya** [comment]
Hey @Tofel I think the PR is fine, although I see a consistent failure in the tests when Ryuk is disabled. Please see https://github.com/testcontainers/testcontainers-go/actions/runs/7299330173/job/19892388225?pr=1971 There are two containers leaked. I thought it was caused for something else, but I see this consistent error build after build.

Could you take a look please?

OTOH, I was planning to include this PR in the next release (today), but finding this makes me postpone it to the following one. Hope you understand.

**Comment by Tofel** [comment]
@mdelapenya sure, totally understand, I'd do the same. Will have a look and fix, but probably won't be able to do it this year :/

**Comment by mdelapenya** [comment]
> but probably won't be able to do it this year

Thankfully the rest of this year is really short 😂 . Thanks for understanding, see you next year 🎄

**Comment by Tofel** [comment]
@mdelapenya should be fixed now, I didn't call `terminateContainerOnEnd()` in my tests, which made containers run indefinitely 

Happy new year 🥂

File: docker.go
Comment (mdelapenya) [inline docker.go:780]:
```suggestion
// GetLogProducerErrorChannel exposes the only way for the consumer to be able to listen to errors and react to them.
func (c *DockerContainer) GetLogProducerErrorChannel() <-chan error {
```

**Review by mdelapenya — COMMENTED** [review]
@Tofel This LGTM, thanks for your time here 👏👏👏

I added a comment regarding the docs for the changes, in two places: the comments for the methods and our docs website. But I'd be fine with including the docs website in a separate PR if you prefer it.

Thanks!

**Comment by Tofel** [comment]
@mdelapenya done! by website did you mean the markdown file you mentioned or something yet different?

File: docs/features/follow_logs.md
Comment (mdelapenya) [inline docs/features/follow_logs.md:61]:
```suggestion
If no parameter is passed a default timeout of 5 seconds will be used. Values below 5 seconds and above 60 seconds will
```

File: docs/features/follow_logs.md
Comment (mdelapenya) [inline docs/features/follow_logs.md:74]:
```suggestion

```golang
```

File: docs/features/follow_logs.md
Comment (mdelapenya) [inline docs/features/follow_logs.md:96]: I cannot suggest here to remove the leading whitespaces (Github UI failing this use case). Could you remove them for L96?

**Review by mdelapenya — COMMENTED** [review]
Thanks for adding the docs, I think it's now clear how to use the new features. I added two comments regarding formatting in the docs. Once fixed, I think this is ready to be merged.

**Comment by Tofel** [comment]
done!

File: docs/features/follow_logs.md
Comment (mdelapenya) [inline docs/features/follow_logs.md:64]: Use H2 instead:
```suggestion
## Listening to errors
```

**Review by mdelapenya — APPROVED** [review]
LGTM, thanks for your work here!

Would you mind checking if the linked issues/PRs are potentially closed?

Cheers!

**Comment by Tofel** [comment]
* https://github.com/testcontainers/testcontainers-go/pull/1278 -- I think we can close this one as it was retrying conecting again in a loop instead of panicking, but currently the same behaviour can be achieved using approach similar to what what I put in the example (listening in a detached goroutine to the error channels, sleeping for 1s if there's an error and starting log producer again)
* https://github.com/testcontainers/testcontainers-go/pull/1911 -- as well as instead of panicking it would go back to beginning immediately and start over 

@mdelapenya: do you have another release planned?

**Pull request merged**
